### PR TITLE
Add Tests For BooleanOption and BaseColumnOptions Classes

### PR DESCRIPTION
**Pull request by sagars729**
- Added tests for BooleanOption Class
- Added class methods for creating a new Options object and also for getting the variable path for an Options object to allow tests to inherit each other (more useful in future tests)
- Added variable path type check to BooleanOption _validate_helper as method and check were not implemented in BaseOption class
- [x] Tests for BaseColumn Class ~~(IP)~~

### CODE REVIEW

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (sagars729) [inline dataprofiler/tests/profilers/test_boolean_option.py:27]: This could be a class variable instead of a method. I wasn't sure if the params of the Option classes may eventually change the variable path.

File: dataprofiler/tests/profilers/test_base_option.py
Comment (AnhTruong) [inline dataprofiler/tests/profilers/test_base_option.py:23]: should name it as `getBaseOption`

Comment (ChrisWallace2020) [reply dataprofiler/tests/profilers/test_base_option.py:23]: But shouldn't that not matter if in the boolean tests you just use super().<test_function>? Because then it'll defer to the super's implementation which can be named something different, and it'll still work in the boolean case? Idk maybe I'm misunderstanding the goal.

Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_option.py:23]: Consider the super().test_validate_helper() call in test_validate_helper for TestBooleanOption. If I used getBaseOption in this class, then regardless of which class is calling super, it will create a new BaseOption object. By creating a getOptions method that is overriden the getOptions call in the TestBaseOption class will create a BooleanOption (when TestBooleanOption is calling its super) which is the intended behavior.

Comment (ChrisWallace2020) [reply dataprofiler/tests/profilers/test_base_option.py:23]: It seems I had a lapse in my python understanding, I thought in the call to super() it wouldn't make calls to the Boolean class, I just tested this out in a jupyter notebook and realized I was wrong. Disregard my comment.

Comment (lettergram) [inline dataprofiler/tests/profilers/test_base_option.py:23]: Why are the tabs 8 spaces here, but only 4 spaces earlier?

Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_option.py:23]: I believe GitHub's default tab spacing is 4. The view will change if you use the following link: [https://github.com/capitalone/DataProfiler/pull/51/files?ts=4](https://github.com/capitalone/DataProfiler/pull/51/files?ts=4)

**Review by lettergram — DISMISSED** [review]

File: dataprofiler/tests/profilers/test_base_option.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_base_option.py:23]: I believe int his case we should use 4 spaces, or set the IDE to do 4 spaces when you hit tab.

Comment (JGSweets) [reply dataprofiler/tests/profilers/test_base_option.py:23]: `setUpClass` is in camel case because of the design of unittest, but i'm not sure `getOptions` should also be camel case but rather snake case as normalized for python.

Comment (JGSweets) [inline dataprofiler/tests/profilers/test_base_option.py:23]: I think it might be better if we use a class variable o define what get options is such that the new tests requires it to be defined. Otherwise, anything that imports this class could forget to alter the getOptions and end up with the wrong unittests. Additionally, it might be best to create it as an abstract test set which then forces the setting.

i.e.
```python

class AbstractTestOptions...(object):
    
    option_class = None
    
    @classmethod
    def get_options(self, *args, **kwargs):
        if not self.option_class:
            raise ....
        return self.option_class(*args, **kwargs)

    .
    .
    . <other abstract tests which are universal>


class TestBaseOptions(AbstractTestOptions, unittest.....):

    option_class = BaseOption
    
    .
    .
    . <Tests etc...>
```

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_boolean_option.py:14]: are these necessary?

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_boolean_option.py:22]: same get options statement

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_boolean_option.py:26]: like getOptions, casing / should be abstract and set as class prop?

Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:26]: should path be inferred by the class name?

**Review by JGSweets — CHANGES_REQUESTED** [review]
Made some initial thoughts.

File: dataprofiler/tests/profilers/test_base_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_option.py:23]: My own IDE does 4 spaces but GitHub does not - I can replace all "\t" with four spaces if that's preferable.

Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_option.py:23]: Sure, I can create an abstract class that is implemented by the other TestOptions classes and have it check and enforce that setting.

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_boolean_option.py:14]: I'm not sure - they were included in the original profiler options tests.

File: dataprofiler/profilers/profiler_options.py
Comment (sagars729) [inline dataprofiler/profilers/profiler_options.py:162]: This is probably redundant since the call to super() will raise the same error

File: dataprofiler/tests/profilers/test_base_option.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_base_option.py:23]: Yes, please use spaces. for the IDE, I believe you can set it to convert your tab to spaces. such that it isn't saved as a \t

File: dataprofiler/profilers/profiler_options.py
Comment (sagars729) [reply dataprofiler/profilers/profiler_options.py:132]: See comment below on repetition in BaseColumnOption. I could add the check to the BaseOption class instead (raise a ValueError before the NotImplementedError) to avoid any repetition.

**Comment by sagars729** [comment]
@JGSweets added new commits that

- Removed redundant type check from BaseColumnOptions class
- Fixed casing of getOptions and getOptionsPath to be get_options and get_options_path
- Replaced all "\t" characters with "    "
- Created a option_class class variable to determine the Class
- Created an abstract test options class that contains get_options and get_options_path methods 
- Inferred option variable path in the get_options_path method rather than explicitly defining it

I am still unsure if the mocks are necessary or not necessary.

File: dataprofiler/tests/profilers/test_base_column_options.py
Comment (sagars729) [inline dataprofiler/tests/profilers/test_base_column_options.py:42]: This is redundant (I am overriding the super method to call the super method). It does make the file easier to read though as it's clear what tests are being run (which is good if these tests need to be modified in the future).

File: dataprofiler/tests/profilers/test_base_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_base_option.py:18]: as above may change depending on how abstractbaseclass is setup

File: dataprofiler/tests/profilers/test_base_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_base_option.py:61]: If these are generic, I would recommend moving them to the abstract base class.

We are treating it more like a mixing for testing TBH.

Comment (JGSweets) [reply dataprofiler/tests/profilers/test_base_option.py:61]: if not generic then fine to keep in place.

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:14]: I don't think they are as we don't use the datalabeler in these tests, however previous tests did have it, e.g. `test_data_labeler`

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_boolean_option.py:75]: I would like weigh in from others, but I think we might want to have option.validate() return None if no errors as opposed to an empty list. Thoughts?

Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:75]: This of course would require updating the validate logic around https://github.com/capitalone/DataProfiler/blob/main/dataprofiler/profilers/profiler_options.py#L109

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_boolean_option.py:23]: we should also assert when setting `is_enabled=True` and `is_enabled=False`

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_boolean_option.py:82]: nice!

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:75]: I only reference here as this is where the api meets the user.

**Review by JGSweets — CHANGES_REQUESTED** [review]
Overall pretty good, some thoughts etc.

File: dataprofiler/profilers/profiler_options.py
Comment (sagars729) [reply dataprofiler/profilers/profiler_options.py:161]: It actually ends up being necessary because of the default argument (variable_path = "BaseColumnOptions") without it, all the variable paths end up being incorrect.

File: dataprofiler/tests/profilers/abstract_test_options.py
Comment (sagars729) [reply dataprofiler/tests/profilers/abstract_test_options.py:3]: I want to avoid adding a setUpClass method since this class isn't inheriting unittest. I will add an option_class variable set it to None.

File: dataprofiler/tests/profilers/abstract_test_options.py
Comment (sagars729) [reply dataprofiler/tests/profilers/abstract_test_options.py:6]: For now, since the files are separate, I can add in that check.

File: dataprofiler/tests/profilers/test_base_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_option.py:61]: They aren't generic, but every test class does need to implement these methods. I will at least have the methods raise a NotImplementedError if they aren't implemented by a test class.

File: dataprofiler/tests/profilers/abstract_test_options.py
Comment (sagars729) [reply dataprofiler/tests/profilers/abstract_test_options.py:4]: Added in new commit

File: dataprofiler/tests/profilers/test_base_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_option.py:58]: added in new commit

File: dataprofiler/tests/profilers/abstract_test_options.py
Comment (sagars729) [reply dataprofiler/tests/profilers/abstract_test_options.py:10]: Added in new commit. I first checked if option_class was of type "type" and then checked if an object of that type was an instance of BaseOption.

File: dataprofiler/profilers/profiler_options.py
Comment (JGSweets) [reply dataprofiler/profilers/profiler_options.py:161]: good point!

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_boolean_option.py:19]: See lines 25-28. Those could be moved to test_init, but I want classes that inherit TestBooleanOption (like TestBaseColumnOptions) to check those (but not call super().test_init() since that would fail).
The purpose of test_init was to verify that default params (which may be assumptions in tests) haven't changed.

File: dataprofiler/tests/profilers/test_base_column_options.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_column_options.py:13]: Which option class? I am a little unclear on what is being referenced

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:19]: I would consider these two separate tests as `_set_helper` should only test the `_set_helper` functionality and not whether the `__init__` properly constructs the option class. I just don't find it as intuitive that init tests would be contained elsewhere.

I think if you want to make the `init` more agnostic, you could do 1 of two things. 

1. only look at `options.is_enabled`
this validates class props are properly being set. We also make the test agnostic to the properties call which likely could be validated elsewhere (perhaps a separate test).
2. setup `test_init` in the abstract test class to take in the expected properties for evaluation

Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:19]: of course other options exist as well, just thinking quickly here.

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_boolean_option.py:71]: My comment on this would be that it may break later implementations of validate_helper. Currently, validate_helper calls its super class to get an initial list of errors then continues to add to that list. if None is returned, you would need to add some different logic:
```python
errors = super()._validate_helper(variable_path=variable_path)
if errors == None: errors = []
...
```

File: dataprofiler/tests/profilers/test_base_column_options.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_base_column_options.py:13]: I guess it isn't necessary to move it, but technically we could get rid of 

```python
@classmethod
def setUpClass(cls):
    cls.option_class = BaseColumnOptions
```

and just do 

```python
class ...:
    option_class = BaseColumnOptions
    .
    .
    .
```

which will be less code in the long run for someone to type

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_boolean_option.py:73]: I wouldn't alter `validate_helper`, only `validate` which isn't referencing itself.

File: dataprofiler/tests/profilers/test_boolean_option.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_boolean_option.py:19]: That does make sense - it isn't intuitive to and init checks in the set_helper method. I modified the code slightly to better fit my intention with the tests. Now, it's still testing enabling and disabling the option with set/set_helper and verifying the init separately in the test_init test.

File: dataprofiler/tests/profilers/test_base_column_options.py
Comment (sagars729) [reply dataprofiler/tests/profilers/test_base_column_options.py:13]: I think that might defeat the purpose of the check in the abstract class. TestBaseColumnOptions would just inherit option_class from BooleanOption and use the BooleanOption class for its tests. If we enforce using the setUpClass method (which is generally included because of unittest), then if TestBaseColumnOptions does not set option_class it would default to None.

File: dataprofiler/profilers/profiler_options.py
Comment (sagars729) [inline dataprofiler/profilers/profiler_options.py:110]: @JGSweets modified to return None on validate with tests changed accordingly

Comment (JGSweets) [reply dataprofiler/profilers/profiler_options.py:110]: ty

**Review by JGSweets — APPROVED** [review]

**Comment by JGSweets** [comment]
@sagars729 please update the brach with master

**Review by AnhTruong — APPROVED** [review]

**Comment by AnhTruong** [comment]
Will merge once the branch is up-to-date with the master

**Comment by AnhTruong** [comment]
@sagars729

**Pull request merged**
